### PR TITLE
mail-filter/amavisd-new: add 2.13.0

### PR DESCRIPTION
**Pull request by matoro**
This version contains much cleanup from upstream and a brand-new test suite.  I don't use this software, so testing would be appreciated from somebody who does.

### CODE REVIEW

File: mail-filter/amavisd-new/amavisd-new-2.13.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+		dev-perl/LWP-Protocol-https
+		dev-perl/Net-SSLeay )
+	selinux? ( sec-policy/selinux-amavis )
+	snmp? ( net-analyzer/net-snmp[perl] )
+	spamassassin? ( mail-filter/spamassassin dev-perl/Image-Info )"
+
+BDEPEND="${RDEPEND}
+	test? (
+		virtual/perl-Test-Harness
+		dev-perl/Test-Class
```
Comment (Dessa) [inline mail-filter/amavisd-new/amavisd-new-2.13.0.ebuild:80]: needs dev-perl/Test-Most too

**Comment by Dessa** [comment]
upstream switched to Dist-Zilla so this needs more work.
i don't think there is a simple way to get this built properly, i took "inspiration" from dev-perl/Lab-Measurement to get it built (yes it builds in its current state, but it will fail to launch since upstream split out some code) but it feels rough around the edges.
i can hopefully find some time this weekend to improve it and provide a patch to keep things moving

**Comment by matoro** [comment]
> upstream switched to Dist-Zilla so this needs more work. i don't think there is a simple way to get this built properly, i took "inspiration" from dev-perl/Lab-Measurement to get it built (yes it builds in its current state, but it will fail to launch since upstream split out some code) but it feels rough around the edges. i can hopefully find some time this weekend to improve it and provide a patch to keep things moving

Thanks!  I saw the Dist-Zilla stuff but from what I read about it, I thought it was more just a way for upstream to easily create releases, and it wasn't necessary for building.

What does it error with when you try to launch it?  Do you have a minimal reproducible command I could use to test it that doesn't require actually setting up a full mail filtering pipeline?

**Comment by Dessa** [comment]
just trying to launch amavisd, without the initscript, should be sufficient:
``Can't locate Amavis/Boot.pm in @INC (you may need to install the Amavis::Boot module) (@INC entries checked: /etc/perl /usr/local/lib64/perl5/5.38/x86_64-linux /usr/local/lib64/perl5/5.38 /usr/lib64/perl5/vendor_perl/5.38/x86_64-linux /usr/lib64/perl5/vendor_perl/5.38 /usr/lib64/perl5/5.38/x86_64-linux /usr/lib64/perl5/5.38 /usr/lib64/perl5/vendor_perl/5.36 /usr/lib64/perl5/vendor_perl/5.34 /usr/lib64/perl5/vendor_perl/5.32 /usr/lib64/perl5/vendor_perl/5.30.3) at /usr/sbin/amavisd line 99.``

**Comment by Dessa** [comment]
alright, i have done stuff (and im mostly not proud of it)
```
--- amavisd-new-2.13.0.ebuild.orig	2023-12-15 22:14:44.058166855 +0100
+++ amavisd-new-2.13.0.ebuild	2023-12-15 22:25:41.444304311 +0100
@@ -1,9 +1,9 @@
 # Copyright 1999-2023 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
-EAPI=7
+EAPI=8
 
-inherit systemd perl-functions
+inherit systemd perl-module
 
 DESCRIPTION="High-performance interface between the MTA and content checkers"
 HOMEPAGE="https://gitlab.com/amavis/amavis"
@@ -11,8 +11,8 @@
 
 LICENSE="GPL-2 BSD-2"
 SLOT="0"
-KEYWORDS="~amd64 ~arm64 ~ppc ~ppc64 ~sparc ~x86"
-IUSE="clamav courier dkim ldap mysql postgres qmail razor rspamd rspamd-https selinux snmp spamassassin test"
+KEYWORDS="~amd64 ~x86"
+IUSE="clamav dkim ldap mysql postgres razor rspamd rspamd-https selinux snmp spamassassin test"
 RESTRICT="!test? ( test )"
 REQUIRED_USE="test? ( spamassassin )"
 
@@ -75,6 +75,8 @@
 	spamassassin? ( mail-filter/spamassassin dev-perl/Image-Info )"
 
 BDEPEND="${RDEPEND}
+	dev-perl/Dist-Zilla
+	virtual/perl-ExtUtils-MakeMaker
 	test? (
 		virtual/perl-Test-Harness
 		dev-perl/Test-Class
@@ -87,14 +89,49 @@
 AMAVIS_ROOT="/var/lib/amavishome"
 S="${WORKDIR}/amavis-v${PV}"
 
-src_prepare() {
-	if use courier ; then
-		eapply -p0 amavisd-new-courier.patch
-	fi
+dzil_to_distdir() {
+	local dzil_root dest has_missing modname dzil_version
+	dzil_root="$1"
+	dest="$2"
+
+	cd "${dzil_root}" || die "Can't enter workdir '${dzil_root}'";
+
+	dzil_version="$(dzil version)" || die "Error invoking 'dzil version'"
+	einfo "Generating CPAN dist with ${dzil_version}"
+
+	has_missing=""
+
+	einfo "Checking dzil authordeps"
+	while IFS= read -d $'\n' -r modname; do
+		if [[ -z "${has_missing}" ]]; then
+			has_missing=1
+			eerror "'dzil authordeps' indicates missing build dependencies"
+			eerror "These will prevent building, please report a bug"
+			eerror "Missing:"
+		fi
+	S=	eerror "  ${modname}"
+	done < <( dzil authordeps --missing --versions )
+
+	[[ -z "${has_missing}" ]] || die "Satisfy all missing authordeps first"
+
+	einfo "Checking dzil build deps"
+	while IFS= read -d $'\n' -r modname; do
+		if [[ -z "${has_missing}" ]]; then
+			has_missing=1
+			ewarn "'dzil listdeps' indicates missing build dependencies"
+			ewarn "These may prevent building, please report a bug if they do"
+			ewarn "Missing:"
+		fi
+		ewarn "  ${modname}"
+	done < <( dzil listdeps --missing --versions --author )
 
-	if use qmail ; then
-		eapply -p0 amavisd-new-qmqpqq.patch
-	fi
+	einfo "Generating release"
+	dzil build --notgz --in "${dest}" || die "Unable to build CPAN dist in '${dest}'"
+}
+
+src_prepare() {
+	# perl-module doesn't account for this being a directory
+	mv README_FILES READ_FILES || die
 
 	# We need to fix the daemon_user and daemon_group in amavis-mc even
 	# though we're going to run it in the foreground, because it calls
@@ -116,11 +153,19 @@
 			"${S}/conf/amavisd.conf" || die "missing conf file - sa"
 	fi
 
-	eapply_user
-
 	# needs ZMQ::LibZMQ3 which only suports net-libs/zeromq-3*,
 	# long since removed from tree
 	perl_rm_files t/Amavis/ZMQTest.t
+	sed -e '/^ZMQ::LibZMQ3 =/d' \
+	-i dist.ini || die "Can't patch dist.ini"
+
+	eapply_user
+
+	# prevent distdir-in-distdir
+	mv "${S}" "${T}" || die
+	dzil_to_distdir "${T}/amavis-v${PV}" "${S}"
+
+	perl-module_src_prepare
 }
 
 src_test() {
@@ -128,14 +173,24 @@
 }
 
 src_install() {
-	dosbin bin/amavisd{,-agent,-nanny,-release,-signer,-status}
-	dobin contrib/p0f-analyzer.pl bin/amavisd-submit
+	perl-module_src_install
+	rm "${ED}"/usr/bin/{amavis-services,amavis-mc,amavisd-snmp-subagent-zmq}
+	mkdir "${ED}"/usr/sbin
+	mv "${ED}"/usr/bin/amavisd "${ED}"/usr/sbin/amavisd || die
+	mv "${ED}"/usr/bin/amavisd-agent "${ED}"/usr/sbin/amavisd-agent || die
+	mv "${ED}"/usr/bin/amavisd-nanny "${ED}"/usr/sbin/amavisd-nanny || die
+	mv "${ED}"/usr/bin/amavisd-release "${ED}"/usr/sbin/amavisd-release || die
+	mv "${ED}"/usr/bin/amavisd-signer "${ED}"/usr/sbin/amavisd-signer || die
+	mv "${ED}"/usr/bin/amavisd-status "${ED}"/usr/sbin/amavisd-status || die
+	dobin contrib/p0f-analyzer.pl
 
 	if use snmp ; then
-		dosbin bin/amavisd-snmp-subagent
+		mv "${ED}"/usr/bin/amavisd-snmp-subagent "${ED}"/usr/sbin/amavisd-snmp-subagent || die
 		newinitd "${FILESDIR}/amavisd-snmp-subagent.initd" \
-				 bin/amavisd-snmp-subagent
+				 amavisd-snmp-subagent
 		dodoc AMAVIS-MIB.txt
+	else
+		rm "${ED}"/usr/bin/amavisd-snmp-subagent
 	fi
 
 	if use ldap ; then
@@ -169,12 +224,12 @@
 		conf/amavisd-docker.conf
 
 	docinto README_FILES
-	dodoc README_FILES/README*
-	dodoc -r README_FILES/*.{html,css}
+	dodoc READ_FILES/README*
+	dodoc -r READ_FILES/*.{html,css}
 	docinto README_FILES/images
-	dodoc README_FILES/images/*.png
+	dodoc READ_FILES/images/*.png
 	docinto README_FILES/images/callouts
-	dodoc README_FILES/images/callouts/*.png
+	dodoc READ_FILES/images/callouts/*.png
 
 	docinto test-messages
 	dodoc t/messages/README
```

- courier and qmail useflags removed, their patches don't apply
- due to Dist-Zilla i had to drop keywords and i hate it.
- yes the newinitd line for snmp was right how it was before ;)
- i hate all the mv'ing tbh.

it works though, i tested most useflags (hence the removal)

**Comment by matoro** [comment]
Wow, that looks great, thanks!  I've pushed it up exactly like that.  Don't worry about keywords, I have all the hardware and can re-add them later.

**Pull request closed without merging**